piker/brokers/kucoin.py

```diff
@@ -0,0 +1,718 @@
+async def open_symbol_search(
+    ctx: tractor.Context,
+) -> None:
+    async with open_cached_client('kucoin') as client:
+        # load all symbols locally for fast search
+        await client.cache_pairs()
+        await ctx.started()
+
+        async with ctx.open_stream() as stream:
+            async for pattern in stream:
+                await stream.send(await client.search_symbols(pattern))
+                log.info('Kucoin symbol search opened')
+
+
+@acm
+async def open_ping_task(
+    ws: wsproto.WSConnection, 
+    ping_interval, connect_id
+) -> AsyncGenerator[None, None]:
+    '''
+    Spawn a non-blocking task that pings the ws
+    server every ping_interval so Kucoin doesn't drop
+    our connection
+
+    '''
+    async with trio.open_nursery() as n:
+        # TODO: cache this task so it's only called once
+        async def ping_server():
+            while True:
+                await trio.sleep((ping_interval - 1000) / 1000)
+                await ws.send_msg({'id': connect_id, 'type': 'ping'})
+
+        log.info('Starting ping task for kucoin ws connection')
+        n.start_soon(ping_server)
+
+        yield
+
+        n.cancel_scope.cancel()
+
+
+async def stream_quotes(
+    send_chan: trio.abc.SendChannel,
+    symbols: list[str],
+    feed_is_live: trio.Event,
+    loglevel: str = '',
+    # startup sync
+    task_status: TaskStatus[tuple[dict, dict]] = trio.TASK_STATUS_IGNORED,
+) -> None:
+    '''
+    Required piker api to stream real-time data.
+    Where the rubber hits the road baby
+
+    '''
+    async with open_cached_client('kucoin') as client:
+        token, ping_interval = await client._get_ws_token()
+        connect_id = str(uuid4())
+        pairs = await client.cache_pairs()
+        ws_url = (
+            f'wss://ws-api-spot.kucoin.com/?'
+            f'token={token}&[connectId={connect_id}]'
+        )
+
+        # open ping task
+        async with (
+            open_autorecon_ws(ws_url) as ws,
+            open_ping_task(ws, ping_interval, connect_id),
+        ):
+            log.info('Starting up quote stream')
+            # loop through symbols and sub to feedz
+            for sym in symbols:
+                pair: KucoinMktPair = pairs[sym]
+                kucoin_sym = pair.symbol
+
+                init_msgs = {
+                    # pass back token, and bool, signalling if we're the writer
+                    # and that history has been written
+                    sym: {
+                        'symbol_info': {
+                            'asset_type': 'crypto',
+                            'price_tick_size': float(pair.baseIncrement),
+                            'lot_tick_size': float(pair.baseMinSize),
+                        },
+                        'shm_write_opts': {'sum_tick_vml': False},
+                        'fqsn': sym,
+                    }
+                }
+
+                async with (
+                    subscribe(ws, connect_id, kucoin_sym),
```

Comment:
FYI, you can also pass this into `open_autorecon_ws(ws_url, fixture=subscribe) as ws,` above. Again nbd we can polish it later especially since that API has been reworked entirely in #489 😂

Reply:
Ah but then I wouldn't have my symbol data.